examples/pytorch/mnist-ddp/mnist.ipynb

```diff
@@ -0,0 +1,627 @@
+    "    from torch.optim.lr_scheduler import StepLR\n",
+    "    from torch.utils.data import DataLoader\n",
+    "    from torch.utils.data.distributed import DistributedSampler\n",
+    "    from torchvision.datasets import FashionMNIST\n",
+    "\n",
+    "    # Define the PyTorch CNN model to be trained\n",
+    "    class Net(nn.Module):\n",
+    "        def __init__(self):\n",
+    "            super(Net, self).__init__()\n",
+    "            self.conv1 = nn.Conv2d(1, 20, 5, 1)\n",
+    "            self.conv2 = nn.Conv2d(20, 50, 5, 1)\n",
+    "            self.fc1 = nn.Linear(4 * 4 * 50, 500)\n",
+    "            self.fc2 = nn.Linear(500, 10)\n",
+    "\n",
+    "        def forward(self, x):\n",
+    "            x = F.relu(self.conv1(x))\n",
+    "            x = F.max_pool2d(x, 2, 2)\n",
+    "            x = F.relu(self.conv2(x))\n",
+    "            x = F.max_pool2d(x, 2, 2)\n",
+    "            x = x.view(-1, 4 * 4 * 50)\n",
+    "            x = F.relu(self.fc1(x))\n",
+    "            x = self.fc2(x)\n",
+    "            return F.log_softmax(x, dim=1)\n",
+    "\n",
+    "    # Use NCCL is a GPU is available, otherwise use Gloo as communication backend\n",
+    "    device, backend = (\"cuda\", \"nccl\") if torch.cuda.is_available() else (\"cpu\", \"gloo\")\n",
+    "\n",
+    "    print(f\"Using Device: {device}, Backend: {backend}\")\n",
+    "\n",
+    "    # Setup PyTorch Distributed\n",
+    "    local_rank = int(os.getenv(\"LOCAL_RANK\", 0))\n",
+    "    dist.init_process_group(backend=backend)\n",
+    "\n",
+    "    print(\n",
+    "        \"Distributed Training for WORLD_SIZE: {}, RANK: {}, LOCAL_RANK: {}\".format(\n",
+    "            dist.get_world_size(),\n",
+    "            dist.get_rank(),\n",
+    "            local_rank,\n",
+    "        )\n",
+    "    )\n",
+    "\n",
+    "    # Create the model and load it into the device\n",
+    "    device = torch.device(f\"{device}:{local_rank}\")\n",
+    "    model = DistributedDataParallel(Net().to(device))\n",
+    "\n",
+    "    # Retrieve the Fashion-MNIST dataset\n",
+    "    dataset = FashionMNIST(\n",
+    "        \"./data\",\n",
+    "        train=True,\n",
+    "        download=True,\n",
+    "        transform=transforms.Compose([transforms.ToTensor()]),\n",
+    "    )\n",
+    "\n",
+    "    # Shard the dataset accross workers\n",
+    "    train_loader = DataLoader(\n",
+    "        dataset,\n",
+    "        batch_size=100,\n",
+    "        sampler=DistributedSampler(dataset),\n",
+    "        pin_memory=torch.cuda.is_available(),\n",
+    "    )\n",
+    "\n",
+    "    # Setup the optimization loop\n",
+    "    criterion = nn.CrossEntropyLoss().to(device)\n",
+    "    optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9)\n",
+    "    scheduler = StepLR(optimizer, step_size=10, gamma=0.8)\n",
+    "\n",
+    "    for epoch in range(1, params.get(\"epochs\", 1) + 1):\n",
```

Comment:
Let's skip the input params from this example for now.
I think, we should create dedicated example for this, especially after we decide on: https://github.com/kubeflow/training-operator/pull/2384